spacelift/resource_bitbucket_datacenter_integration.go

```diff
@@ -0,0 +1,213 @@
+package spacelift
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal"
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal/structs"
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal/validations"
+)
+
+const (
+	bitbucketDatacenterAccessToken = "access_token"
+)
+
+func resourceBitbucketDatacenterIntegration() *schema.Resource {
+	return &schema.Resource{
+		Description: "`spacelift_bitbucket_datacenter_integration` represents details of a bitbucket datacenter integration",
+
+		CreateContext: resourceBitbucketDatacenterIntegrationCreate,
+		ReadContext:   resourceBitbucketDatacenterIntegrationRead,
+		UpdateContext: resourceBitbucketDatacenterIntegrationUpdate,
+		DeleteContext: resourceBitbucketDatacenterIntegrationDelete,
+
+		Schema: map[string]*schema.Schema{
+			bitbucketDatacenterID: {
+				Type:        schema.TypeString,
+				Description: "Bitbucket Datacenter integration id.",
+				Computed:    true,
+			},
+			bitbucketDatacenterName: {
+				Type:        schema.TypeString,
+				Description: "Bitbucket Datacenter integration name",
+				Required:    true,
+			},
+			bitbucketDatacenterIsDefault: {
+				Type:        schema.TypeBool,
+				Description: "Bitbucket Datacenter integration is default.",
+				Required:    true,
+			},
+			bitbucketDatacenterSpaceID: {
+				Type:        schema.TypeString,
+				Description: "Bitbucket Datacenter integration space id. Defaults to `root`.",
+				Optional:    true,
+				Computed:    true,
+			},
+			bitbucketDatacenterLabels: {
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type:             schema.TypeString,
+					ValidateDiagFunc: validations.DisallowEmptyString,
+				},
+				Description: "Bitbucket Datacenter integration labels",
+				Optional:    true,
+				Computed:    true,
+			},
+			bitbucketDatacenterDescription: {
+				Type:        schema.TypeString,
+				Description: "Bitbucket Datacenter integration description",
+				Optional:    true,
+				Computed:    true,
+			},
+			bitbucketDatacenterAPIHost: {
+				Type:        schema.TypeString,
+				Description: "The API host where requests will be sent",
+				Required:    true,
+			},
+			bitbucketDatacenterUserFacingHost: {
+				Type:        schema.TypeString,
+				Description: "User Facing Host which will be user for all user-facing URLs displayed in the Spacelift UI",
+				Required:    true,
```

Comment:
Actually, it would be hard. @adamconnelly 

For example, we've set up integration with defined `api_host` and `user_facing_host`. 
Then we set up `api_host` only. 
We expect `user_facing_host` to be changed to the `api_host` value. 

But it's not. Because we also fill value from the server (it's computed). In `*schema.ResourceData`, we have an old value, not an empty one. We could use `Default`, but the value isn't constant (depends on `api_host` value).

Maybe there is a workaround for that. For now, I leave it as it is.

```go
testSteps(t, []resource.TestStep{
			{
				Config: `
					resource "spacelift_bitbucket_datacenter_integration" "test" {
						name              = "` + name + `"
						is_default        = false
						space_id          = "root"
						api_host          = "` + host + `"
						username          = "username"
						access_token      = "access-token" 
					}
				`,
				Check: Resource(
					resourceName,
					Attribute("api_host", Equals(host)),
					Attribute("user_facing_host", Equals(host)),
				),
			},
			{
				Config: `
					resource "spacelift_bitbucket_datacenter_integration" "test" {
						name              = "` + name + `"
						is_default        = false
						space_id          = "root"
						api_host          = "` + host + `"
						user_facing_host  = "https://something-else"
						username          = "username"
						access_token      = "access-token" 
					}
				`,
				Check: Resource(
					resourceName,
					Attribute("api_host", Equals(host)),
					Attribute("user_facing_host", Equals("https://something-else")),
				),
			},
			{
				Config: `
					resource "spacelift_bitbucket_datacenter_integration" "test" {
						name              = "` + name + `"
						is_default        = false
						space_id          = "root"
						api_host          = "` + host + `"
						// removed: user_facing_host  = "https://something-else"
						username          = "username"
						access_token      = "access-token" 
					}
				`,
				Check: Resource(
					resourceName,
					Attribute("api_host", Equals(host)),
					Attribute("user_facing_host", Equals(host)),
				),
			},
		})
```

```
test_provider.go:30: Step 3/3 error: Check failed: check 2 on resource spacelift_bitbucket_datacenter_integration.test failed: expected "https://bitbucket.com/ejjj9", got "https://something-else" instead
```

Reply:
No worries.